src/misc/lv_array.c

```diff
@@ -45,71 +45,143 @@ void lv_array_init(lv_array_t * array, uint32_t capacity, uint32_t element_size)
 }
 
-void lv_array_destroy(lv_array_t * array)
+void lv_array_deinit(lv_array_t * array)
 {
     if(array->data) {
         lv_free(array->data);
         array->data = NULL;
     }
+
+    array->size = 0;
+    array->capacity = 0;
 }
 
-void lv_array_copy(lv_array_t * target, const lv_array_t * array)
+void lv_array_copy(lv_array_t * target, const lv_array_t * source)
 {
-    if(lv_array_is_empty(array)) {
+    if(lv_array_is_empty(source)) {
         return;
     }
-    lv_array_destroy(target);
-    lv_array_init(target, array->capacity, array->element_size);
-    lv_memcpy(target->data, array->data, array->size * array->element_size);
-    target->size = array->size;
+    lv_array_deinit(target);
+    lv_array_init(target, source->capacity, source->element_size);
+    lv_memcpy(target->data, source->data, source->size * source->element_size);
+    target->size = source->size;
 }
 
 void lv_array_clear(lv_array_t * array)
 {
     array->size = 0;
 }
 
-void lv_array_resize(lv_array_t * array, uint32_t new_capacity)
+/**
+ * Remove the element at the specified position in the array.
+ * @param array pointer to an `lv_array_t` variable
+ * @param index the index of the element to remove
+ */
+bool lv_array_remove(lv_array_t * array, uint32_t index)
+{
+    if(index >= array->size) {
+        return false;
+    }
+
+    /*Shortcut*/
+    if(index == array->size - 1) {
+        lv_array_resize(array, array->size - 1);
+        return false;
+    }
+
+    uint8_t * start = lv_array_at(array, index);
+    uint8_t * remaining = start + array->element_size;
+    uint32_t remaining_size = (array->size - index - 1) * array->element_size;
+    lv_memcpy(start, remaining, remaining_size);
+    lv_array_resize(array, array->size - 1);
+    return true;
+}
+
+bool lv_array_erase(lv_array_t * array, uint32_t start, uint32_t end)
 {
-    if(new_capacity > array->size) {
-        if(new_capacity > array->capacity) {
-            uint8_t * data = lv_malloc(new_capacity * array->element_size);
-            lv_memcpy(data, array->data, array->size * array->element_size);
-            lv_free(array->data);
-            array->data = data;
-            array->capacity = new_capacity;
-        }
+    if(end > array->size) {
+        end = array->size;
+    }
+
+    if(start >= end) {
+        return false;
+    }
+
+    /*Shortcut*/
+    if(end == array->size) {
+        lv_array_resize(array, start);
+        return true;
     }
-    else {
+
+    uint8_t * start_p = lv_array_at(array, start);
+    uint8_t * remaining = start_p + (end - start) * array->element_size;
+    uint32_t remaining_size = (array->size - end) * array->element_size;
+    lv_memcpy(start_p, remaining, remaining_size);
+    lv_array_resize(array, array->size - (end - start));
+    return true;
+}
+
+void lv_array_resize(lv_array_t * array, uint32_t new_capacity)
+{
+    uint8_t * data = lv_realloc(array->data, new_capacity * array->element_size);
+    LV_ASSERT_NULL(data);
```

Comment:
In case `lv_array_remove(array, 0)` is called (removing last item in array), it invokes `lv_array_resize(array, 0)`, and so `lv_realloc()` will return `NULL`, and this function will fail.
Please correct me if I'm wrong.
There could be also a test case for adding then removing an item.

Reply:
Sorry, in case `lv_realloc()` is called with size 0, and , then in actual LVGL implementation it will return the address of `memory_zero` and not `NULL`:
1. https://github.com/lvgl/lvgl/blob/master/src/stdlib/lv_mem.c#L131
2. https://github.com/lvgl/lvgl/blob/master/src/stdlib/lv_mem.c#L26
3. https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h#L85

... and so `array->data` will point to `memory_zero`, and adding new item will overwrite anything in `lv_global_t` struct.
Or am I wrong?

Reply:
>  in case lv_realloc() is called with size 0, then in actual LVGL implementation it will return the address of memory_zero and not NULL

I confirm it. It's because allocating e.g. a zero sized array means that the pointer is used and initialized but it happened to have 0 size. While NULL means an error, not set, etc. Similarly to Python or JS:
```py
>>> a = []
>>> a
[]
>>> b = None
>>> b
>>> 
```